pkg/workflow/dataimpl.go

```diff
@@ -11,62 +14,158 @@ import (
 	Content_location_key string = "Content-Location"
 )
 
-// NewDataFromInput creates a new data instance from the given input data.
-//
-// It will preserve the headers, metadata and errors
-func NewDataFromInput(input Data, typeIdentifier Identifier, contentType string, payload interface{}) Data {
-	if len(typeIdentifier.Path) <= 0 {
-		panic("Given identifier is not a type identifier")
+type Option = func(d *DataImpl)
+
+func withIdentifier(id *Identifier) Option {
+	return func(d *DataImpl) {
+		d.identifier = *id
+	}
+}
+
+func withContentType(contentType *string) Option {
+	return func(d *DataImpl) {
+		d.contentType = *contentType
+	}
+}
+
+func withPayload(payload *interface{}) Option {
+	return func(d *DataImpl) {
+		d.payload = *payload
+	}
+}
+
+func WithConfiguration(config *configuration.Configuration) Option {
+	return func(d *DataImpl) {
+		d.configuration = *config
+	}
+}
+
+func WithInputData(input Data) Option {
+	return func(d *DataImpl) {
+		d.inputData = input
+	}
+}
+
+func WithLogger(logger *zerolog.Logger) Option {
+	return func(d *DataImpl) {
+		d.logger = logger
+	}
+}
+
+func newDataWith(opts ...Option) Data {
+	// initial DataImpl
+	output := &DataImpl{
+		payloadLocation: Location{
+			Path: "",
+			Type: InMemory,
+		},
+		configuration: configuration.NewInMemory(),
+		logger:        &zerolog.Logger{},
+		inputData:     nil,
+	}
+	for _, opt := range opts {
+		opt(output)
 	}
 
-	dataIdentifier := *typeIdentifier
-	dataIdentifier.Scheme = "did"
+	logger := *output.logger
 
-	header := http.Header{
-		Content_type_key: {contentType},
+	// validate DataImpl
+	if len(output.identifier.Path) <= 0 {
+		panic("Given identifier is not a type identifier")
 	}
 
-	var errors []snyk_errors.Error
+	// update DataImpl values
+	output.identifier.Scheme = "did"
+
+	output.header = http.Header{
+		Content_type_key: {output.contentType},
+	}
 
-	if input != nil {
+	if output.inputData != nil {
 		// derive fragment from input data if available
-		dataIdentifier.Fragment = input.GetIdentifier().Fragment
+		output.identifier.Fragment = output.inputData.GetIdentifier().Fragment
 
 		// derive content location from input
-		if loc, err := input.GetMetaData(Content_location_key); err == nil {
-			header.Add(Content_location_key, loc)
+		if loc, err := output.inputData.GetMetaData(Content_location_key); err == nil {
+			output.header.Add(Content_location_key, loc)
 		}
 
-		errors = slices.Clone(input.GetErrorList())
+		output.errors = slices.Clone(output.inputData.GetErrorList())
 	} else {
 		// generate time based fragment
-		dataIdentifier.Fragment = fmt.Sprintf("%d", time.Now().Nanosecond())
+		output.identifier.Fragment = fmt.Sprintf("%d", time.Now().Nanosecond())
 	}
 
-	output := &DataImpl{
-		identifier: &dataIdentifier,
-		header:     header,
-		payload:    payload,
-		errors:     errors,
+	output.payloadLocation = setPayloadLocation(output.identifier, output.configuration, output.payload, output.logger)
```

Comment:
Suggestion: keep logic close together, separating writing to disk and related logic makes it difficult to understand and error prone. Currently there are two places that use `setPayloadLocation()` which can easily cause the errors. Can't we just call a common function? that does the same in both cases?